File: py-joerg/board.py
import random
from collections import defaultdict
from enum import Enum
from typing import List, Deque, Optional, Dict, Set, Callable

from cards.card import Card
from order import Order
from played_card import PlayedCard
from player import Player
from log import new_logger

LOGGER = new_logger("board")


class PlayerStates(Enum):
    HandFaceDown = 1
    UnableToWin = 2


class Victory(Exception):
    pass


class ActivePlayerState:
    def __init__(self, state: PlayerStates, duration: int, action: Callable[[], None]):
        self.state: PlayerStates = state
        # 0 denotes indefinite.
        self.duration: int = duration
        self.action: Callable[[], None] = action


class PlayerStateHandler:
    def __init__(self):
        self.states: Dict[PlayerStates, ActivePlayerState] = {}

    def add_state(self, state, duration, action):
        self.states[state] = ActivePlayerState(state, duration, action)

    def has_state(self, state: PlayerStates) -> bool:
        return state in self.states

    def remove(self, ps: PlayerStates) -> None:
        assert ps in self.states
        self.states[ps].action()
        del self.states[ps]

    def update(self) -> None:
        cleared_states = set()
        for state in self.states:
            if self.states[state].duration <= 0:
                cleared_states.add(state)
            self.states[state].duration -= 1

        for state_to_remove in cleared_states:
            self.remove(state_to_remove)


class Board:
    def __init__(
        self,
        deck: Deque[Card],
        number_of_players: int,
        starting_hand_size: int,
        wins_needed: int,
    ):
        self.pole: Player
        self.cube: List[Card] = list(deck)
        self.deck: Deque[Card] = deck
        self.played_cards: List[PlayedCard] = []
        self.players: List[Player] = []
        self.round_winning_card: Optional[Card] = None
        self.round_winner: Optional[Player] = None
        self.graveyard: Dict[Player, List[Card]] = defaultdict(list)
        self.victories: Dict[Player, int] = defaultdict(int)

        # Card blocked for _int_ number of turns.
        self.blocked_cards: Dict[Card, int] = defaultdict(int)
        self.player_states: Dict[Player, PlayerStateHandler] = {}

        self.starting_hand_size: int = starting_hand_size
        self.wins_needed: int = wins_needed

        for p in range(number_of_players):
            player = Player(p)
            self.players.append(player)
            self.graveyard[player] = []
            self.victories[player] = 0
            self.player_states[player] = PlayerStateHandler()

    def randomly_assign_pole(self) -> Player:
        pole_player = self.get_random_player()
        self.set_pole(pole_player)
        return pole_player

    def set_pole(self, player: Player) -> None:
        self.pole = player

    def shuffle_deck(self) -> None:
        random.shuffle(self.deck)

    def commit_phase(self) -> None:
        for player in self.players:
            valid_cards = self.valid_plays(player.hand)
            assert len(valid_cards) >= 1, f"No valid cards to play! {player.hand}"

            chosen_card = player_picks(player, valid_cards, "Chose a card to commit!")
            player.remove_card_from_hand(chosen_card)

            chosen_order = player_picks(player, [Order.attack, Order.defense], "Chose card order!")
            self.commit_card(player, chosen_card, chosen_order)

    def progress_pole(self) -> None:
        next_player = self.get_next_player(self.get_pole_player())
        self.set_pole(next_player)

    def resolve_on_reveal(self) -> None:
        for resolving_card in self.get_played_cards():
            resolving_card.revealed = True
            resolving_card.on_reveal(self)

    def resolve_before_power(self) -> None:
        for resolving_card in self.get_played_cards():
            resolving_card.before_power(self)

    def resolve_win_lose(self) -> None:
        for resolving_card in self.get_played_cards():
            if resolving_card.card == self.round_winning_card:
                resolving_card.on_win(self)
            elif resolving_card.card != self.round_winning_card:
                resolving_card.on_lose(self)

    def end_resolve_phase(self) -> None:
        # TODO(_): Bug with boar last round.
        if self.round_winner:
            self.add_to_graveyard(self.round_winner, self.round_winning_card)

        self.return_losing_cards()
        # Remove all played cards on board.
        self.played_cards = []

    def return_losing_cards(self) -> None:
        for played_card in self.losing_cards():
            played_card.player.add_card_to_hand(played_card.card)

    def update_blocked_cards(self) -> None:
        cleared_cards: Set[Card] = set()
        for c in self.blocked_cards:
            if self.blocked_cards[c] <= 0:
                cleared_cards.add(c)
            self.blocked_cards[c] -= 1

        for c in cleared_cards:
            del self.blocked_cards[c]

    def update_player_states(self) -> None:
        for player_states in self.player_states.values():
            player_states.update()

    def begin_round(self) -> None:
        # Flush old cards.
        self.played_cards = []
        self.round_winner = None
        self.round_winning_card = None

        for p in self.players:
            assert self.starting_hand_size - p.hand_size() == len(
                self.graveyard[p]
            ), f"Player does not have the right number of cards on hand! Starting hand size: {self.starting_hand_size} | {p} Hand: {p.hand} | Graveyard: {self.graveyard[p]} | "

    def end_round(self) -> None:
        self.update_blocked_cards()
        self.update_player_states()

    def commit_card(self, player: Player, card: Card, order: Order):
        if card.legendary:
            self.blocked_cards[card] = 1
        self.played_cards.append(PlayedCard(player, card, order))

    def number_of_players(self) -> int:
        return len(self.players)

    def valid_plays(self, cards: List[Card]) -> List[Card]:
        valid_cards = [c for c in cards if c not in self.blocked_cards]
        return valid_cards

    def get_players_played_card(self, player: Player) -> PlayedCard:
        for ac in self.played_cards:
            if ac.player != player:
                continue

            return ac

        raise RuntimeError("Player haven't played a card this round!")

    def get_previous_player(self, player: Player) -> Player:
        origin_player_index = self.players.index(player)
        return self.players[origin_player_index - 1 % self.number_of_players()]

    def get_played_cards(self):
        start_index = self.get_pole_index()
        for i in range(self.number_of_players()):
            yield self.played_cards[(start_index + i) % len(self.played_cards)]

    def deal_cards(self) -> None:
        for player in self.players:
            for _ in range(self.starting_hand_size):
                card = self.deck.pop()
                player.add_card_to_hand(card)

        assert player.hand_size() == self.starting_hand_size

    def draw_card(self) -> Card:
        # TODO(_): Undefined behavior when deck is empty.
        return self.deck.popleft()

    def get_opponents(self, player: Player) -> List[Player]:
        return [p for p in self.players if p != player]

    def get_random_player(self) -> Player:
        return random.choice(self.players)

    def get_random_opponent(self, player: Player) -> Player:
        return random.choice(self.get_opponents(player))

    def cycle_event(self, triggering_player: Player) -> None:
        falling_behind_players = self.all_players_except_winner(triggering_player)

        cycled_cards = []
        for player in falling_behind_players:
            cycled_card = self.cycle_for_player(player)
            cycled_cards.append(cycled_card)
        self.add_cycled_cards_to_bottom_of_deck(cycled_cards)

    def cycle_for_player(self, player: Player) -> Card:
        chosen_card = player.player_picks(player.hand, "Chose card to cycle.")

        self.player_cycle_card(player, chosen_card)

        return chosen_card

    def player_cycle_card(self, player: Player, card: Card):
        player.remove_card_from_hand(card)
        card.on_cycle(self, player)

    def put_card_at_bottom_of_deck(self, card: Card) -> None:
        self.deck.append(card)

    def players_in_pole_order_from_player(self, player: Player) -> List[Player]:
        start_index = self.players.index(player)
        for i in range(self.number_of_players()):
            yield self.players[(i + start_index) % self.number_of_players()]

    def get_pole_index(self) -> int:
        return next(
            index
            for index, ac in enumerate(self.played_cards)
            if ac.player == self.pole
        )

    def get_pole_card(self) -> PlayedCard:
        pole_index = self.get_pole_index()
        return self.played_cards[pole_index]

    def all_players_except_winner(self, winning_player: Player) -> List[Player]:
        return [p for p in self.players if p != winning_player]

    def resolved_order(self) -> Order:
        orders = [c.order for c in self.played_cards]
        num_attack = 0
        num_defense = 0
        for order in orders:
            if order == Order.attack:
                num_attack += 1
            elif order == Order.defense:
                num_defense += 1

        assert num_attack + num_defense == len(self.played_cards)

        # Tie
        if num_attack > num_defense:
            return Order.attack
        elif num_defense > num_attack:
            return Order.defense
        elif num_attack == num_defense:
            pole_played_card = self.get_pole_card()
            return pole_played_card.order

        raise RuntimeError("Unreachable")

    def get_pole_player(self) -> Player:
        return self.pole

    def get_opponent_played_cards(self, player: Player) -> List[PlayedCard]:
        return [ac for ac in self.get_played_cards() if ac.player != player]

    def get_next_player(self, player: Player) -> Player:
        start_index = self.players.index(player)
        return self.players[(start_index + 1) % self.number_of_players()]

    def get_card_index(self, card: Card) -> int:
        return next(idx for idx, ac in enumerate(self.played_cards) if ac.card == card)

    def swap_ownage_of_played_cards(self, card1: Card, card2: Card) -> None:
        index_card1 = self.get_card_index(card1)
        index_card2 = self.get_card_index(card2)

        tmp_card = self.played_cards[index_card1].card
        self.played_cards[index_card1].card = self.played_cards[index_card2].card
        self.played_cards[index_card2].card = tmp_card

        self.played_cards[index_card1]

    def cycle_phase(self):
        if self.round_winner and self.player_will_win_next_round(self.round_winner):
            LOGGER.info(f"Cycle! {self.round_winner} has reached 2 wins.")
            LOGGER.info(" ")
            self.cycle_event(self.round_winner)

    def resolve_power(self) -> PlayedCard:
        best_card = None

        def is_better(a: Card, b: Card):
            a.power_resolve()
            b.power_resolve()
            # Is card _a_ better than card _b_?
            if self.resolved_order() == Order.defense:
                # Lower is better.
                return a.power < b.power
            else:
                # Higher is better.
                return a.power > b.power

        for played_card in self.get_played_cards():
            if not best_card or is_better(played_card.card, best_card.card):
                best_card = played_card

        if not best_card:
            raise RuntimeError(f"Unable to find best card! Board: {self}")

        self.best_card = best_card

    def player_will_win_next_round(self, player: Player) -> bool:
        return self.victories[player] + 1 == self.wins_needed

    def resolve_winner(self) -> None:
        if self.player_will_win_next_round(
            self.best_card.player
        ) and self.player_states[self.best_card.player].has_state(
            PlayerStates.UnableToWin
        ):
            return

        self.set_round_winner(self.best_card.player)
        self.set_round_winning_card(self.best_card.card.copy())
        self.victories[self.best_card.player] += 1

        LOGGER.info("")
        LOGGER.info("")
        LOGGER.info(
            f"Winning card! {self.best_card.card} played by {self.best_card.player}"
        )
        if (
            self.best_card.player
            and self.victories[self.best_card.player] == self.wins_needed
        ):
            raise Victory()

    def add_to_graveyard(self, player: Player, card: Card) -> None:
        self.graveyard[player].append(card)

    def set_round_winner(self, player: Player) -> None:
        self.round_winner = player

    def set_round_winning_card(self, card: Card) -> None:
        self.round_winning_card = card

    def is_losing_card(self, card: Card) -> bool:
        return self.round_winning_card != card

    def is_opponent_card(self, card: Card, player: Player) -> bool:
        ac = next(c for c in self.played_cards if c.card == card)
        return ac.player != player

    def add_cycled_cards_to_bottom_of_deck(self, cycled_cards: List[Card]):
        self.deck.extend(cycled_cards)

    def losing_cards(self) -> List[PlayedCard]:
        all_cards = self.played_cards
        return [c for c in all_cards if c.card != self.round_winning_card]

    def trade(self, player1: Player, card1: Card, player2: Player, card2: Card) -> None:
        assert card1 in player1.hand
        assert card2 in player2.hand

        index1, _ = player1.remove_card_from_hand(card1)
        index2, _ = player2.remove_card_from_hand(card2)

        player1.add_card_to_hand(card2, index=index1)
        player2.add_card_to_hand(card1, index=index2)

        assert card2 in player1.hand
        assert card1 in player2.hand

    def serialize(self):
        return {
            "pole": self.pole.num,
            "cube": self.cube,
            "deck": self.deck,
            "played_cards": self.played_cards,
            "players": self.players,
            "round_winning_card": self.round_winning_card,
            "round_winner": self.round_winner,
            "graveyard": {p.num: graveyard for p, graveyard in self.graveyard.items()},
            "victories": {p.num: victories for p, victories in self.victories.items()},
            # Card blocked for _int_ number of turns.
            "blocked_cards": {c.name: turns for c, turns in self.blocked_cards.items()},
            "player_states": {p.num: {asdf.name: qwerty for asdf, qwerty in ps.states.items()} for p, ps in self.player_states.items()},
            "starting_hand_size": self.starting_hand_size,
            "wins_needed": self.wins_needed,
        }

    def __repr__(self):
        return f"Pole: {self.pole}\nResolved order: {self.resolved_order()}"

File: joergo/board.go
package joerg

import (
	"errors"
	"fmt"
	"log"
	"math/rand"
)

const (
	LIBRARY_PATH    = "../cards.json"
	MINIMUM_PLAYERS = 3
	MAXIMUM_PLAYERS = 6
)

// https://play.golang.org/p/i1BGlRsP19
type Board struct {
	Players          []*Player    `json:"players"`
	Pole             *Player      `json:"pole"`
	Cube             *Cube        `json:"cube"`
	Deck             []Carder     `json:"deck"` // Make private later.
	PlayedCards      []PlayedCard `json:"playedCards"`
	RoundWinner      *Player      `json:"roundWinner"`
	RoundWinningCard Carder       `json:"roundWinningCard"`
	NumPlayers       uint         `json:"numPlayers"`
	StartingHandSize uint         `json:"startingHandSize"`
	WinsNeeded       uint         `json:"winsNeeded"`
	BestCard         *PlayedCard  `json:"bestCard"` // Make private later.
	Victories        map[int]int  `json:"victories"`
}

func NewBoard(
	startingHandSize uint,
	winsNeeded uint,
	players []*Player,
) (*Board, error) {
	numPlayers := len(players)
	cube, err := ReadCube(LIBRARY_PATH)
	if err != nil {
		return nil, err
	}

	if numPlayers < MINIMUM_PLAYERS || numPlayers > MAXIMUM_PLAYERS {
		return nil, errors.New("illegal number of numPlayers")
	}
	if winsNeeded == 0 || winsNeeded > startingHandSize {
		return nil, errors.New("illegal winsNeeded")
	}
	if int(startingHandSize)*numPlayers > len(cube.Cards)-int(numPlayers) {
		return nil, errors.New("illegal startingHandSize")
	}

	victories := make(map[int]int)
	for _, p := range players {
		victories[p.Num] = 0
	}

	b := Board{
		Players:          players,
		NumPlayers:       uint(numPlayers),
		StartingHandSize: startingHandSize,
		WinsNeeded:       winsNeeded,
		Cube:             cube,
		Deck:             cube.Cards,
		Victories:        victories,
	}
	return &b, nil
}

func (b *Board) AddPlayer(p *Player) (err error) {
	if len(b.Players) == int(b.NumPlayers) {
		return errors.New("unable to add any more players(!)")
	}
	b.Players = append(b.Players, p)
	return nil
}

func (b *Board) AddBotPlayer() (err error) {
	return nil
}

func (b *Board) Next() {
	fmt.Println("Board next called!")
}

func (b *Board) RandomlyAssignPole() {
	b.Pole = b.Players[rand.Intn(len(b.Players))]
}

func (b *Board) ShuffleDeck() {
	rand.Shuffle(len(b.Deck), func(i, j int) {
		b.Deck[i], b.Deck[j] = b.Deck[j], b.Deck[i]
	})
}

func (b *Board) DealCards() {
	for _, p := range b.Players {
		p.Hand = b.Deck[:b.StartingHandSize]
		b.Deck = b.Deck[b.StartingHandSize:]
	}
}

func (b *Board) ResolveOnReveal() {
	log.Println("ResolveOnReveal")
	for _, pc := range b.PlayedCards {
		pc.Revealed = true
		pc.Card.OnReveal(b, pc.Player)
		log.Println(pc.Card)
	}
}

func (b *Board) ResolveWinLose() {
	log.Println("ResolveOnReveal")
	for _, pc := range b.PlayedCards {
		if pc.Card.Name() == b.RoundWinningCard.Name() {
			pc.Card.OnWin()
		} else {
			pc.Card.OnLose()
		}
	}
}

func (b *Board) EndResolvePhase() {
	b.ReturnLosingCards()
	b.PlayedCards = []PlayedCard{}
}

func (b *Board) LosingCards() (pcs []PlayedCard) {
	for _, pc := range b.PlayedCards {
		if pc.Card.Name() == b.RoundWinningCard.Name() {
			continue
		}
		pcs = append(pcs, pc)
	}
	return pcs
}

func (b *Board) ReturnLosingCards() {
	for _, pc := range b.LosingCards() {
		pc.Player.AddCardToHand(pc.Card)
	}
}

func (b *Board) AllPlayersExceptWinner() (ps []*Player) {
	for _, p := range b.Players {
		if p.Name == b.RoundWinner.Name {
			continue
		}
		ps = append(ps, p)
	}
	return ps
}

func (b *Board) AddCycledCardsToBottomOfDeck(cycledCards []Carder) {
	b.Deck = append(b.Deck, cycledCards...)
}

func (b *Board) cycleEvent(trgP *Player) error {
	fallingBehind := b.AllPlayersExceptWinner()

	cycledCards := []Carder{}
	for _, p := range fallingBehind {
		cycledCard, err := b.cycleForPlayer(p)
		if err != nil {
			return err
		}
		cycledCards = append(cycledCards, cycledCard)
	}
	b.AddCycledCardsToBottomOfDeck(cycledCards)
	return nil
}

func (b *Board) cycleForPlayer(p *Player) (Carder, error) {
	selection := make([]interface{}, 0, len(p.Hand))
	for _, c := range p.Hand {
		selection = append(selection, c)
	}
	chosenCardInt, err := p.Picks(selection, "Chose card to cycle.", 1)
	if err != nil {
		return nil, err
	}
	chosenCard, ok := chosenCardInt.(Carder)
	if !ok {
		return nil, errors.New("unable to type assert cycled card")
	}

	p.RemoveCardFromHand(chosenCard)

	newCard := b.DrawCard()
	p.AddCardToHand(newCard)

	return chosenCard, nil
}

func (b *Board) DrawCard() Carder {
	if len(b.Deck) < 1 {
		panic("DrawCard: no cards to draw!")
	}
	tmp := b.Deck[0]
	b.Deck = b.Deck[1:]
	return tmp
}

// def cycle_for_player(self, player: Player) -> Card:
//     chosen_card = player_picks(player.hand, "Chose card to cycle.")

//     self.player_cycle_card(player, chosen_card)

//     return chosen_card

// def player_cycle_card(self, player: Player, card: Card):
//     player.remove_card_from_hand(card)
//     card.on_cycle(self, player)

func (b *Board) PlayerWillWinNextRound(p *Player) bool {
	return b.Victories[p.Num] == int(b.WinsNeeded)-1
}

func (b *Board) CyclePhase() error {
	if b.RoundWinner != nil && b.PlayerWillWinNextRound(b.RoundWinner) {
		if err := b.cycleEvent(b.RoundWinner); err != nil {
			return err
		}
	}
	return nil
}

func (b *Board) ProgressPole() {
	b.Pole = b.NextPlayer(b.Pole)
}

func (b *Board) NextPlayer(p *Player) *Player {
	return b.Players[(p.Num+1)%len(b.Players)]
}

func (b *Board) EndRound() {
	// Update blocked cards.
	// Update player statuses.
}

func (b *Board) ResolveBeforePower() {
	log.Println("ResolveBeforePower")
	for _, pc := range b.PlayedCards {
		log.Println(pc)
	}
}

func (b *Board) PoleCard() PlayedCard {
	fmt.Println(b.PlayedCards)
	return b.PlayedCards[b.Pole.Num]
}

func (b *Board) ResolveOrder() Order {
	atk, def := 0, 0
	for _, pc := range b.PlayedCards {
		if pc.Order == OrderAttack {
			atk += 1
		} else if pc.Order == OrderDefense {
			def += 1
		}
	}

	if atk+def != len(b.PlayedCards) {
		log.Println("Something went wrong!")
	}

	if atk > def {
		return OrderAttack
	} else if def > atk {
		return OrderDefense
	} else if atk == def {
		polePc := b.PoleCard()
		return polePc.Order
	}
	panic("Unreachable")
}

func (b *Board) PlayedCardsInOrder() []PlayedCard {
	cards := make([]PlayedCard, 0, len(b.Players))
	for i := 0; i < len(b.Players); i++ {
		c := b.PlayedCards[(i+b.Pole.Num)%len(b.Players)]
		cards = append(cards, c)
	}
	return cards
}

func (b *Board) ResolvePower() {
	var bestCard *PlayedCard
	log.Println("ResolvePower")
	defCmp := func(a Carder, b Carder) bool {
		return a.Power() > b.Power()
	}
	atkCmp := func(a Carder, b Carder) bool {
		return a.Power() > b.Power()
	}
	var isBetter func(Carder, Carder) bool
	resolvedOrder := b.ResolveOrder()
	if resolvedOrder == OrderAttack {
		// Higher is better.
		isBetter = atkCmp
	} else {
		// Lower is better.
		isBetter = defCmp
	}

	for _, pc := range b.PlayedCardsInOrder() {
		if bestCard == nil || isBetter(pc.Card, bestCard.Card) {
			bestCard = &pc
		}
	}
	if bestCard == nil {
		panic("Unable to find best card!")
	}

	b.BestCard = bestCard
}

func (b *Board) ResolveWinner() {
	// if self.player_will_win_next_round(
	//     self.best_card.player
	// ) and self.player_states[self.best_card.player].has_state(
	//     PlayerStates.UnableToWin
	// ):
	//     return

	b.RoundWinner = b.BestCard.Player
	b.RoundWinningCard = b.BestCard.Card
	b.Victories[b.RoundWinner.Num] += 1

	log.Println("Winning card!", b.RoundWinningCard.Name())
	if b.Victories[b.RoundWinner.Num] == int(b.WinsNeeded) {
		panic(fmt.Sprintf("Player %s wins the game!", b.RoundWinner.Name))
	}
}

func (b *Board) BeginRound() {
	b.PlayedCards = make([]PlayedCard, 0, len(b.Players))
	b.RoundWinner = nil
	b.RoundWinningCard = nil
}

func (b *Board) commitCard(p *Player, c Carder, o Order) {
	pc := PlayedCard{
		Player:   p,
		Card:     c,
		Order:    o,
		Revealed: false,
	}
	b.PlayedCards = append(b.PlayedCards, pc)
}

func (b *Board) CommitPhase() (err error) {
	for _, p := range b.Players {
		selection := make([]interface{}, 0, len(p.Hand))
		for _, c := range p.Hand {
			selection = append(selection, c)
		}
		chosenCardInt, err := p.Picks(selection, p.Name+": Choose a card to commit!", 1)
		if err != nil {
			return err
		}
		chosenCard, ok := chosenCardInt.(Carder)
		if !ok {
			return errors.New("Type assertion failed")
		}

		p.RemoveCardFromHand(chosenCard)

		selection = make([]interface{}, 0, 2)
		selection = append(selection, OrderAttack)
		selection = append(selection, OrderDefense)
		chosenOrderInt, err := p.Picks(selection, p.Name+": Choose card's order!", 1)
		if err != nil {
			return err
		}
		chosenOrder, ok := chosenOrderInt.(Order)
		if !ok {
			return errors.New("Type assertion failed")
		}
		fmt.Println("Chosen card: \n", chosenCard)
		fmt.Printf("Chosen card: %#v\n", chosenCard)
		fmt.Printf("Chosen order: %s\n", chosenOrder)
		b.commitCard(p, chosenCard, chosenOrder)
	}
	return nil
}
